rpc/grpc/api.go

```diff
@@ -37,3 +41,149 @@ func (bapi *broadcastAPI) BroadcastTx(ctx context.Context, req *RequestBroadcast
 		},
 	}, nil
 }
+
+type blockAPI struct {
+	sync.Mutex
+	heightListeners map[chan int64]struct{}
+}
+
+func newBlockAPI() *blockAPI {
+	return &blockAPI{
+		// TODO(rach-id) make 1000 configurable if there is a need for it
+		heightListeners: make(map[chan int64]struct{}, 1000),
+	}
+}
+
+func (blockAPI *blockAPI) listenForHeights() {
+	env := core.GetEnvironment()
+	newBlocksChannel, err := env.EventBus.Subscribe(context.Background(), "grpc-listener", types2.EventQueryNewBlock, 500)
+	if err != nil {
+		log.Fatalf("Failed to subscribe to new blocks: %v", err)
+	}
+	for {
+		select {
+		case <-newBlocksChannel.Cancelled():
+			env.Logger.Error("cancelled grpc subscription")
+			// TODO(rach-id): maybe retry to connect if users want this functionality
+			return
+		case event := <-newBlocksChannel.Out():
+			newBlockEvent, ok := event.Events()[types2.EventTypeKey]
+			if !ok || len(newBlockEvent) == 0 || newBlockEvent[0] != types2.EventNewBlock {
+				continue
+			}
+			data, ok := event.Data().(types2.EventDataNewBlock)
+			if !ok {
+				env.Logger.Debug("couldn't cast event data to new block")
+				continue
+			}
+			blockAPI.broadcastToListeners(data.Block.Height)
+		}
+	}
+
+}
+
+func (blockAPI *blockAPI) broadcastToListeners(height int64) {
+	for ch := range blockAPI.heightListeners {
+		select {
+		case ch <- height:
+		default:
+		}
+	}
+}
+
+func (blockAPI *blockAPI) addHeightListener() chan int64 {
+	blockAPI.Lock()
+	defer blockAPI.Unlock()
+	ch := make(chan int64, 50)
+	blockAPI.heightListeners[ch] = struct{}{}
+	return ch
+}
+
+func (blockAPI *blockAPI) removeHeightListener(ch chan int64) {
+	blockAPI.Lock()
+	defer blockAPI.Unlock()
+	delete(blockAPI.heightListeners, ch)
+	close(ch)
+}
+
+func (blockAPI *blockAPI) BlockByHash(req *BlockByHashRequest, stream BlockAPI_BlockByHashServer) error {
+	// Question: do you want just the block, or you want the response to be similar to what
+	// the RPC endpoint returns: a ResultBlock?
+	block, err := core.GetEnvironment().BlockStore.LoadBlockByHash(req.Hash).ToProto()
+	if err != nil {
+		return err
+	}
+	err = stream.Send(&BlockByHashResponse{Block: block})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (blockAPI *blockAPI) BlockByHeight(req *BlockByHeightRequest, stream BlockAPI_BlockByHeightServer) error {
+	block, err := core.GetEnvironment().BlockStore.LoadBlock(req.Height).ToProto()
+	if err != nil {
+		return err
+	}
+	err = stream.Send(&BlockByHeightResponse{
+		Block: block,
+	})
```

Comment:
Looks like rsponse is not explicitly batched. Does it work well for large blocks (>100mb)?

Reply:
my bad, this is the streaming part :D 65d936db01c1c48bb2a64bc82294d85d9c00d1d8

forgot it locally

Reply:
example usage:

```
	blockStreamer, err := client.BlockByHeight(context.Background(), &coregrpc.BlockByHeightRequest{Height: 2})
	if err != nil {
		panic(err)
	}
	blockMeta, err := client.BlockMetaByHeight(context.Background(), &coregrpc.BlockMetaByHeightRequest{Height: 2})
	if err != nil {
		panic(err)
	}
	parts := make([]*core.Part, 0)
	for i := 0; i < int(blockMeta.BlockMeta.BlockID.PartSetHeader.Total); i++ {
		resp, err := blockStreamer.Recv()
		if err != nil {
			panic(err)
		}
		parts = append(parts, resp.BlockPart)
		if resp.IsLast && i < int(blockMeta.BlockMeta.BlockID.PartSetHeader.Total)-1 {
			panic("couldn't get all parts")
		} else if resp.IsLast {
			break
		}
	}

	h := types.NewPartSetFromHeader(types.PartSetHeader{
		Total: blockMeta.BlockMeta.BlockID.PartSetHeader.Total,
		Hash:  blockMeta.BlockMeta.BlockID.PartSetHeader.Hash,
	})

	for _, part := range parts {
		ok, err := h.AddPart(&types.Part{
			Index: part.Index,
			Bytes: part.Bytes,
			Proof: merkle.Proof{
				Total:    part.Proof.Total,
				Index:    part.Proof.Index,
				LeafHash: part.Proof.LeafHash,
				Aunts:    part.Proof.Aunts,
			},
		})
		if err != nil {
			panic(err)
		}
		if !ok {
			panic("not okey")
		}
	}
	pbb := new(core.Block)
	bz, err := io.ReadAll(h.GetReader())
	if err != nil {
		panic(err)
	}
	err = proto.Unmarshal(bz, pbb)
	if err != nil {
		panic(err)
	}
	block, err := types.BlockFromProto(pbb)
	if err != nil {
		panic(err)
	}
	fmt.Println(block)
```

Reply:
Looks like parts are constructed with proofs and other supporting information that would allow them to be individually used. Is it used for efficient streaming of chunks from core side? 
Would it be possible to send Block as chunks of constructed eds? This should be probably most efficient usage of bandwidth for data transfer.

Reply:
> Is it used for efficient streaming of chunks from core side?

they're used so individual chunks can be authenticated to avoid receiving invalid ones.

> Would it be possible to send Block as chunks of constructed eds?

The EDS is not stored in core. This means that we will need to erasure encode the block before sending it. And in this case, I guess it's better if it's erasure encoded at the destination

Reply:
> they're used so individual chunks can be authenticated to avoid receiving invalid ones.

In case of bridge node connected to consensus node, this is not needed, because most of the time it is trusted connection. May we add the option to have only data in response to reduce proofs overhead? 

>The EDS is not stored in core. This means that we will need to erasure encode the block before sending it. And in this case, I guess it's better if it's erasure encoded at the destination

Sorry, was about to suggest to stream only ODS part. And only the part with data omitting the padding? This should be the most compact representation of block, thus most efficient for transfer.

Reply:
>  May we add the option to have only data in response to reduce proofs overhead?

sure

> to stream only ODS part

only ODS is being streamed now.

Reply:
[49befbd](https://github.com/celestiaorg/celestia-core/pull/1513/commits/49befbdf688c65eb8a52f6780832f68f33c173c2)

Reply:
4304c44af6973292546f0c419ed1109c58ca1160